File: Data Structure I/3. Linked List/P21.cpp
// Date: Sat, Oct 2, 2021
// 21. Merge Two Sorted Lists (https://leetcode.com/problems/merge-two-sorted-lists/)
// Redo

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

// recursive approach
// time complexity: O(m+n)
// space complexity: O(1)
ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
	if (!l1 || !l2)
		return l1 ? l1 : l2;
	if (l1->val == l2->val) {
		ListNode* result = mergeTwoLists(l1->next, l2->next);
		l1->next = l2;
		l2->next = result;
		return l1;
	}
	if (l1->val < l2->val) {
		l1->next = mergeTwoLists(l1->next, l2);
		return l1;
	}
	l2->next = mergeTwoLists(l1, l2->next);
	return l2;
}

File: LeetCoding-Challenge/March-2022/P21.java
package leetcodingchallenge.march2022;

/**
 * Date: March 07, 2022
 * 21. Merge Two Sorted Lists
 * https://leetcode.com/problems/merge-two-sorted-lists/
 */

public class P21 {
    // time: O(m + n)
    // space: O(1)
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null || list2 == null) return list1 == null ? list2 : list1;
        if (list1.val < list2.val) {
            list1.next = mergeTwoLists(list1.next, list2);
            return list1;
        }
        list2.next = mergeTwoLists(list1, list2.next);
        return list2;
    }

    // Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}
